collector/lib/NetworkStatusInspector.cpp

```diff
@@ -0,0 +1,177 @@
+const std::string NetworkStatusInspector::kEndpointRoute = kBaseRoute + "/endpoint";
+const std::string NetworkStatusInspector::kConnectionRoute = kBaseRoute + "/connection";
+
+const std::string NetworkStatusInspector::kQueryParam_container = "container";
+
+NetworkStatusInspector::NetworkStatusInspector(const std::shared_ptr<ConnectionTracker> conntracker) : conntracker_(conntracker) {
+}
+
+static std::string IP2str(const Address& addr) {
+  switch (addr.family()) {
+    case Address::Family::IPV4: {
+      char str[INET_ADDRSTRLEN];
+      inet_ntop(AF_INET, addr.data(), str, INET_ADDRSTRLEN);
+      return str;
+    }
+    case Address::Family::IPV6: {
+      char str[INET6_ADDRSTRLEN];
+      inet_ntop(AF_INET6, addr.data(), str, INET6_ADDRSTRLEN);
+      return str;
+    }
+    default:
+      break;
+  }
+  return "unk";
+}
+
+static std::string Proto2str(L4Proto proto) {
+  switch (proto) {
+    case collector::L4Proto::TCP:
+      return "TCP";
+    case collector::L4Proto::UDP:
+      return "UDP";
+    case collector::L4Proto::ICMP:
+      return "ICMP";
+    default:
+      break;
+  }
+  return "UNKNOWN";
+}
+
+static std::string IPNet2str(const IPNet& net) {
+  return IP2str(net.address()) + (net.bits() == 0 ? "" : "/" + std::to_string(net.bits()));
+}
+
+static void SerializeEndpoint(const Endpoint& ep, Json::Value& node) {
+  node["address"] = IPNet2str(ep.network());
+  node["port"] = ep.port();
+}
+
+bool NetworkStatusInspector::handleGetEndpoints(struct mg_connection* conn, const QueryParams& queryParams) {
+  collector::AdvertisedEndpointMap endpointStates = conntracker_->FetchEndpointState(true, false);
+  Json::Value bodyRoot(Json::objectValue);
+
+  std::unordered_map<std::string, std::forward_list<collector::ContainerEndpointMap::value_type*>> byContainer;
+
+  std::optional<std::string> containerFilter = GetParameter(queryParams, kQueryParam_container);
+
+  for (auto& endpointState : endpointStates) {
+    if (!containerFilter || *containerFilter == endpointState.first.container())
+      byContainer[endpointState.first.container()].push_front(&endpointState);
+  }
+
+  for (auto& container : byContainer) {
+    const std::string& containerId = container.first;
+
+    Json::Value containerArray(Json::arrayValue);
+
+    for (auto endpointState : container.second) {
+      const collector::ContainerEndpoint& endpoint = endpointState->first;
+      const collector::ConnStatus& status = endpointState->second;
+
+      Json::Value endpointNode(Json::objectValue);
+
+      endpointNode["active"] = status.IsActive();
+
+      endpointNode["l4proto"] = Proto2str(endpoint.l4proto());
+
+      endpointNode["endpoint"] = Json::objectValue;
+      SerializeEndpoint(endpoint.endpoint(), endpointNode["endpoint"]);
+
+      containerArray.append(endpointNode);
+    }
+
+    bodyRoot[containerId] = containerArray;
+  }
+
+  std::string buffer = Json::writeString(jsonStreamWriterBuilder_, bodyRoot);
```

Comment:
Haven't had time to test this, do you know if the output of `Json::writeString` is formatted? I prefer it unformatted, since I pipe it into `jq` most of the times anyways, but wither way works I guess.